lib/cudnn/nnlib.jl

```diff
@@ -13,259 +13,225 @@ const CUDNNFloat = Union{Float16,Float32,Float64}
 
 # Since CUDNN does not support 1D convolution, Conv in Flux will give a CUDNNError if the size is 1-dimensional.
 fix1d(x) = x
-fix1d(x::DenseCuArray{T, 3}) where T = reshape(x, size(x, 1), 1, size(x, 2), size(x, 3))
+fix1d(x::DenseCuArray{T, 3}) where T = reshape(x, 1, size(x, 1), size(x, 2), size(x, 3))
 fix1d(cdims::DenseConvDims{1,K,C_in,C_out,S,P,D,F}) where {K,C_in,C_out,S,P,D,F} =
-  DenseConvDims{2,(K...,1),C_in,C_out,(S...,1),(P...,0,0),(D...,1),F}((cdims.I...,1))
+    DenseConvDims{2,(1,K...),C_in,C_out,(1,S...),(0,0,P...),(1,D...),F}((1,cdims.I...))
 fix1d(pdims::PoolDims{1,K,S,P,D}) where {K,S,P,D,F} =
-  PoolDims{2,(K...,1),(S...,1),(P...,0,0),(D...,1)}((pdims.I..., 1), pdims.C_in)
-
-# We have to reshape the CuArray/PoolDims/DenseConvDims to 4D before feeding to CUDNN.
-reshape4D(x::AbstractVector) = reshape(x, 1, 1, length(x), 1)
-reshape4D(x::AbstractMatrix) = reshape(x, 1, 1, size(x)...)
-reshape4D(x::AbstractArray{T,3}) where T = reshape(x, size(x, 1), 1, size(x, 2), size(x, 3))
-reshape4D(x::AbstractArray{T}) where T = x
-
-workspacesize(x) = min(Mem.info()[1] ÷ 16, sizeof(x) * 2)
-
-function perfChoose(perfResults, returnedAlgoCount)::UInt32
-  if perfResults[1].status != 0
-    return 0
-  else
-    (best_algo,best_time,best_memory) = (perfResults[1].algo,perfResults[1].time,perfResults[1].memory)
-    for i = 2:returnedAlgoCount
-      if perfResults[i].status == 0 && perfResults[i].memory < best_memory && perfResults[i].time < best_time * 1.1
-        (best_algo,best_memory) = (perfResults[i].algo,perfResults[i].memory)
-      end
-    end
-    return best_algo
-  end
-end
-
+    PoolDims{2,(1,K...),(1,S...),(0,0,P...),(1,D...)}((1,pdims.I...), pdims.C_in)
 
 # Softmax
 
-# in-place for x or dy
+# @denizyuret: do not do inplace operations when (1) cpu version is not, (2) one can use softmax!
+# @denizyuret: recalculating y in ∇softmax! is a big waste, the nnlib API should be changed.
 softmax(x::DenseCuArray{T}; dims=1) where T<:CUDNNFloat =
```

Comment:
could we remove this `softmax` ?
This CUDNN implementation is inconsistent with NNlib.jl's `softmax`
so the results of CPU's and GPU's are different.

I just filed a bug: https://github.com/JuliaGPU/CUDA.jl/issues/599